lib/base/type.cpp

```diff
@@ -32,6 +38,60 @@ INITIALIZE_ONCE_WITH_PRIORITY([]() {
 	Type::Register(type);
 }, InitializePriority::RegisterTypeType);
 
+static std::vector<Type::Ptr> l_SortedByLoadDependencies;
+static Atomic l_SortingByLoadDependenciesDone (false);
+
+typedef std::unordered_map<Type*, bool> Visited; // https://stackoverflow.com/a/8942986
+
+INITIALIZE_ONCE_WITH_PRIORITY([] {
+	auto types (Type::GetAllTypes());
+
+	types.erase(std::remove_if(types.begin(), types.end(), [](auto& type) {
+		return !ConfigObject::TypeInstance->IsAssignableFrom(type);
+	}), types.end());
+
+	// Depth-first search
+	std::unordered_set<Type*> unsorted;
+	Visited visited;
+	std::vector<Type::Ptr> sorted;
+
+	for (auto type : types) {
+		unsorted.emplace(type.get());
+	}
+
+	std::function<void(Type*)> visit;
+	visit = ([&visit, &unsorted, &visited, &sorted](Type* type) {
+		if (unsorted.find(type) == unsorted.end()) {
+			return;
+		}
+
+		bool& alreadyVisited (visited.at(type));
+		VERIFY(!alreadyVisited);
+		alreadyVisited = true;
+
+		for (auto dep : type->GetLoadDependencies()) {
+			visit(dep);
+		}
+
+		unsorted.erase(type);
+		sorted.emplace_back(type);
+	});
+
+	while (!unsorted.empty()) {
+		for (auto& type : types) {
+			visited[type.get()] = false;
+		}
+
+		visit(*unsorted.begin());
+	}
+
+	VERIFY(sorted.size() == types.size());
+	VERIFY(sorted[0]->GetLoadDependencies().empty());
```

Comment:
Looks like a nice sanity check, but (obviously) just checks the first element. What about a test case for the whole order? Like iterate the whole vector and for each element, check that it's dependencies are placed at a lower index.

Reply:
> What about a test case for the whole order?

This is unnecessary IMHO, given that if the orders get messed up, you won't be able to start Icinga 2 anyway, so the startup phase actually takes care of such sanity checks and that function should just do what it is meant to do.

Reply:
> the startup phase actually takes care of such sanity

Does it check that explicitly? Or is it just that there's a good chance you'll get some errors then (but maybe it could just be broken in a subtle way)?

Reply:
No, not explicitly, but the type orders do matter when starting Icinga 2, and if you insist on introducing such explicit sanity checks, then unittest it, but this function should not be bothered in checking the final result.

Reply:
> What about a test case

I was thinking of a unit test when writing that.

Reply:
> > What about a test case
> 
> I was thinking of a unit test when writing that.

I thought you'd want to loop through the final result at the end and check their orders!